ihmc-common-walking-control-modules/src/main/java/us/ihmc/commonWalkingControlModules/controlModules/foot/WorkspaceLimiterControlModule.java

```diff
@@ -466,31 +478,27 @@ private void correctHeightForOverExtensionForSingularityAvoidance()
 
       isSwingSingularityAvoidanceUsedOnHeight.set(true);
 
-      updateFractionOfSingularityAvoidanceToUse(alphaSwingSingularityAvoidanceForHeight,
+      updateFractionOfSingularityAvoidanceToUse(desiredPercentOfLegLengthSwing.getDoubleValue(),
+                                                alphaSwingSingularityAvoidanceForHeight,
                                                 maxPercentOfLegLengthForSingularityAvoidanceInSwingForHeight.getValue(),
                                                 percentOfLegLengthMarginToEnableSingularityAvoidanceForHeight.getValue());
 
-      double desiredFootPositionInAxisFrame = -Math.min(desiredLegLength.getDoubleValue(),
-                                                        maxPercentOfLegLengthForSingularityAvoidanceInSwingForHeight.getValue()
-                                                                                           * maximumLegLength.getDoubleValue());
+      double lengthToStartLimiting =
+            (maxPercentOfLegLengthForSingularityAvoidanceInSwingForHeight.getValue() - percentOfLegLengthMarginToEnableSingularityAvoidanceForHeight.getValue())
+            * maximumLegLength.getDoubleValue();
+      double maxLengthToAllow = maxPercentOfLegLengthForSingularityAvoidanceInSwingForHeight.getValue() * maximumLegLength.getDoubleValue();
+      double lengthToUse = InterpolationTools.linearInterpolate(lengthToStartLimiting,
+                                                                 maxLengthToAllow,
+                                                                 alphaSwingSingularityAvoidanceForHeight.getDoubleValue());
 
-      // Mix the desired leg extension velocity to progressively follow the pelvis velocity as the the leg is more straight
-      double desiredLinearVelocityZ = InterpolationTools.linearInterpolate(desiredFootLinearVelocity.getZ(),
-                                                                           pelvisLinearVelocity.getZ(),
-                                                                           alphaSwingSingularityAvoidanceForHeight.getDoubleValue());
+      double desiredFootPositionInAxisFrame = -Math.min(desiredLegLengthSwing.getDoubleValue(), lengthToUse);
```

Comment:
As it was implemented, this would not actually do any limiting until the full limit amount had been reached (that is, until `alphaSwingSingularityAvoidanceForHeight` reached 1.0). Now, this has the limiting phase in, so that it is a lot smoother.